File: spider/spider.go
package spider

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"runtime"
	"strings"

	"github.com/DiheChen/go-asmr-spider/config"
	"github.com/DiheChen/go-asmr-spider/utils"
)

var conf = config.GetConfig()

type OAuth2 struct {
	Authorization string
}

type track struct {
	Type             string  `json:"type"`
	Title            string  `json:"title"`
	Children         []track `json:"children,omitempty"`
	Hash             string  `json:"hash,omitempty"`
	WorkTitle        string  `json:"workTitle,omitempty"`
	MediaStreamURL   string  `json:"mediaStreamUrl,omitempty"`
	MediaDownloadURL string  `json:"mediaDownloadUrl,omitempty"`
}

func Login() (*OAuth2, error) {
	payload, err := json.Marshal(map[string]string{
		"name":     conf.Account,
		"password": conf.Password,
	})
	if err != nil {
		fmt.Println("登录失败, 配置文件有误。")
		return nil, err
	}
	client := utils.Client.Get().(*http.Client)
	req, _ := http.NewRequest("POST", "https://api.asmr.one/api/auth/me", bytes.NewBuffer(payload))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Referer", "https://www.asmr.one/")
	req.Header.Set("User-Agent", "PostmanRuntime/7.29.0")
	resp, err := client.Do(req)
	utils.Client.Put(client)
	if err != nil {
		fmt.Println("登录失败, 网络错误。请尝试通过环境变量的方式设置代理。")
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	all, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Print("登录失败, 读取响应失败。")
		return nil, err
	}
	res := make(map[string]string)
	err = json.Unmarshal(all, &res)
	return &OAuth2{Authorization: "Bearer " + res["token"]}, nil
}

func (OAuth2 *OAuth2) GetVoiceTracks(id string) ([]track, error) {
	client := utils.Client.Get().(*http.Client)
	req, _ := http.NewRequest("GET", "https://api.asmr.one/api/tracks/"+id, nil)
	req.Header.Set("Authorization", OAuth2.Authorization)
	req.Header.Set("Referer", "https://www.asmr.one/")
	req.Header.Set("User-Agent", "PostmanRuntime/7.29.0")
	resp, err := client.Do(req)
	utils.Client.Put(client)
	if err != nil {
		fmt.Println("获取音声信息失败:", err)
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	all, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("获取音声信息失败: ", err)
		return nil, err
	}
	res := make([]track, 0)
	err = json.Unmarshal(all, &res)
	return res, nil
}

func DownloadFile(url string, dirPath string, fileName string) {
	if runtime.GOOS == "windows" {
		for _, str := range []string{"?", "<", ">", ":", "/", "\\", "*", "|"} {
			fileName = strings.Replace(fileName, str, "_", -1)
		}
	}
	savePath := dirPath + "/" + fileName
	if pathExists(savePath) {
		fmt.Println("文件已存在, 跳过。")
		return
	}
	client := utils.Client.Get().(*http.Client)
	fmt.Println("正在下载 " + savePath)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Referer", "https://www.asmr.one/")
	req.Header.Set("User-Agent", "PostmanRuntime/7.29.0")
	resp, err := client.Do(req)
	utils.Client.Put(client)
	if err != nil {
		fmt.Println("下载"+savePath+"失败: ", err)
		return
	}
	defer func() { _ = resp.Body.Close() }()
	file, err := os.OpenFile(dirPath+"/"+fileName+".temp", os.O_CREATE|os.O_APPEND|os.O_RDWR, 0666)
	if err != nil {
		fmt.Println("下载"+savePath+"失败: ", err)
		return
	}
	_, err = io.Copy(file, resp.Body)
	if err != nil {
		fmt.Println("写入"+savePath+"失败: ", err)
		return
	}
	_ = file.Close()
	err = os.Rename(dirPath+"/"+fileName+".temp", dirPath+"/"+fileName)
	if err != nil {
		fmt.Println("重命名"+savePath+"失败: ", err)
		return
	}
	return
}

func EnsureDir(tracks []track, basePath string) {
	path := basePath
	_ = os.MkdirAll(path, os.ModePerm)
	for _, t := range tracks {
		if t.Type != "folder" {
			DownloadFile(t.MediaDownloadURL, path, t.Title)
		} else {
			EnsureDir(t.Children, fmt.Sprintf("%s/%s", path, t.Title))
		}
	}
}

func pathExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil || errors.Is(err, os.ErrExist)
}

func (OAuth2 *OAuth2) Download(id string) {
	id = strings.Replace(id, "RJ", "", 1)
	fmt.Println("作品 RJ 号: " + id)
	tracks, err := OAuth2.GetVoiceTracks(id)
	if err != nil {
		fmt.Println("获取作品失败: " + err.Error())
		return
	}
	EnsureDir(tracks, "RJ"+id)
	fmt.Println("下载完成。")
}

File: app/spider.py
"""
 - Author       : DiheChen
 - Date         : 2022-05-14 07:26:07
 - LastEditors  : DiheChen
 - LastEditTime : 2022-05-14 17:49:48
 - Description  : None
 - GitHub       : https://github.com/DiheChen
"""
from os import makedirs, path
from sys import stdout
from typing import Any, Dict, List

try:
    import ujson as json
except ImportError:
    import json

from aiohttp import ClientSession
from loguru import logger

from .config import config

makedirs("logs", exist_ok=True)
logger.remove()
format_ = "<g>{time:MM-DD HH:mm:ss}</g> [<lvl>{level}</lvl>] | {message}"
logger.add(stdout, format=format_)
logger.add(
    "logs/spider.log",
    format=format_,
    rotation="1 day",
)


class ASMRSpider:
    def __init__(self, name: str, password: str) -> None:
        self.name = name
        self.password = password
        self.headers = {
            "Referer": "https://www.asmr.one/",
            "User-Agent": "PostmanRuntime/7.29.0",
        }

    async def login(self) -> None:
        async with self._session.post(
            "https://api.asmr.one/api/auth/me",
            json={"name": self.name, "password": self.password},
            headers=self.headers,
            proxy=config.proxy,
        ) as resp:
            self.headers |= {
                "Authorization": f"Bearer {(await resp.json())['token']}",
            }

    async def get_voice_info(self, voice_id: str) -> Dict[str, Any]:
        async with self._session.get(
            f"https://api.asmr.one/api/work/{voice_id}",
            headers=self.headers,
            proxy=config.proxy,
        ) as resp:
            return await resp.json()

    async def get_voice_tracks(self, voice_id):
        async with self._session.get(
            f"https://api.asmr.one/api/tracks/{voice_id}",
            headers=self.headers,
            proxy=config.proxy,
        ) as resp:
            return await resp.json()

    async def download_file(self, url: str, save_path: str, file_name: str) -> None:
        file_name = file_name.translate(str.maketrans('/\:*?"<>|', "_________"))
        file_path = path.join(save_path, file_name)
        if not path.exists(file_path):
            logger.info(f"Downloading {file_path}")
            async with self._session.get(
                url, headers=self.headers, proxy=config.proxy, timeout=114514
            ) as resp:
                with open(file_path, "wb") as f:
                    f.write(await resp.read())

    async def ensure_dir(self, tracks: List[Dict[str, Any]], root: str) -> None:
        root_path = root
        folders = [track for track in tracks if track["type"] == "folder"]
        files = [track for track in tracks if track["type"] != "folder"]
        for file in files:
            try:
                await self.download_file(
                    file["mediaDownloadUrl"], root_path, file["title"]
                )
            except Exception as e:
                logger.error(e)
                continue
        for folder in folders:
            new_path = path.join(root_path, folder["title"])
            makedirs(new_path, exist_ok=True)
            await self.ensure_dir(folder["children"], new_path)

    async def download(self, voice_id: str) -> None:
        voice_id = voice_id.strip().split("RJ")[-1]
        voice_info = await self.get_voice_info(voice_id)
        for key in (
            "has_subtitle",
            "create_date",
            "userRating",
            "review_text",
            "progress",
            "updated_at",
            "user_name",
        ):
            voice_info.pop(key)
        root = path.join("Voice", f"RJ{voice_id}")
        makedirs(root, exist_ok=True)
        with open(path.join(root, f"RJ{voice_id}.json"), "w", encoding="utf-8") as f:
            json.dump(voice_info, f, ensure_ascii=False, indent=4)
        tracks = await self.get_voice_tracks(voice_id)
        await self.ensure_dir(tracks, root)

    async def __aenter__(self) -> "ASMRSpider":
        self._session = ClientSession()
        await self.login()
        return self

    async def __aexit__(self, *args) -> None:
        await self._session.close()
